en/tutorials-and-examples/cms/authentication-plugin.rst

```diff
@@ -0,0 +1,287 @@
+CMS Tutorial - Authentication using cakephp/authentication
+##########################################################
+
+CakePHP now supports Authentication and Authorization features using the
+core plugins cakephp/authentication and cakephp/authorization.
+
+In this tutorial we are going to ensure passwords are stored securely in
+our database, we are going to provide a working login page, and apply some
+basic access control to the article creation & editing experiences.
+
+We will assume you have created the ``Controller``, ``Table``, ``Entity`` and
+templates for the ``users`` table in your database, you can do this manually
+like you did before for the ArticlesController, or you can use the bake shell
+to generate the classes for you using ``bin/cake bake all users``.
+
+Adding Password Hashing
+-----------------------
+
+If you were to create/update a user at this point in time, you might notice that
+the passwords are stored in plain text. This is really bad from a security point
+of view, so lets fix that.
+
+This is also a good time to talk about the model layer in CakePHP. In CakePHP,
+we separate the methods that operate on a collection of objects, and a single
+object into different classes. Methods that operate on the collection of
+entities are put in the ``Table`` class, while features belonging to a single
+record are put on the ``Entity`` class.
+
+For example, password hashing is done on the individual record, so we'll
+implement this behavior on the entity object. Because we want to hash the
+password each time it is set, we'll use a mutator/setter method. CakePHP will
+call a convention based setter method any time a property is set in one of your
+entities. Let's add a setter for the password. In **src/Model/Entity/User.php**
+add the following::
+
+    <?php
+    namespace App\Model\Entity;
+
+    use Authentication\PasswordHasher\DefaultPasswordHasher; // Add this line
+    use Cake\ORM\Entity;
+
+    class User extends Entity
+    {
+
+        // Code from bake.
+
+        // Add this method
+        protected function _setPassword(string $password) : ?string
+        {
+            if (strlen($password) > 0) {
+                return (new DefaultPasswordHasher())->hash($password);
+            }
+        }
+    }
+
+Now, point your browser to **http://localhost:8765/users** to see a list of users.
+Remember you'll need to have your local server running, start a standalone PHP
+server using ``bin/cake server``
+
+You can edit the default user that was created during
+:doc:`Installation <installation>`. If you change that user's password,
+you should see a hashed password instead of the original value on the list or
+view pages. CakePHP hashes passwords with `bcrypt
+<http://codahale.com/how-to-safely-store-a-password/>`_ by default. We recommend
+bcrypt for all new applications to keep your security standards high.
```

Comment:
Might want to mention it is the default for PHP as well (with password_hash) to suggest it's more than just our arbitrary pick.